packages/model/src/type-system/implementations/union.ts

```diff
@@ -86,19 +90,49 @@ class UnionTypeImpl<Ts extends types.Types> extends DefaultMethods<types.UnionTy
     value: unknown,
     decodingOptions?: decoding.Options,
   ): decoding.Result<types.Infer<types.UnionType<Ts>>> {
-    if (typeof value === 'object' && value) {
-      const object = value as Record<string, any>
-      const variantName = singleKeyFromObject(object)
-      if (variantName !== undefined && Object.keys(this.variants).includes(variantName)) {
-        return types
-          .concretise(this.variants[variantName])
-          .decodeWithoutValidation(object[variantName], decodingOptions)
-          .map((value) => Object.fromEntries([[variantName, value]]) as types.Infer<types.UnionType<Ts>>)
-          .mapError((errors) => prependVariantToAll(errors, variantName))
+    if (this.isTaggedUnion()) {
+      if (typeof value === 'object' && value) {
+        const object = value as Record<string, any>
+        const variantName = singleKeyFromObject(object)
+        if (variantName !== undefined && Object.keys(this.variants).includes(variantName)) {
+          return types
+            .concretise(this.variants[variantName])
+            .decodeWithoutValidation(object[variantName], decodingOptions)
+            .map((value) => Object.fromEntries([[variantName, value]]) as types.Infer<types.UnionType<Ts>>)
+            .mapError((errors) => prependVariantToAll(errors, variantName))
+        }
+      }
+      const prettyVariants = Object.keys(this.variants).join(' | ')
+      return decoding.fail(`union (${prettyVariants})`, value)
+    } else {
+      const errors: decoding.Error[] = []
+      let potentialDecoded: types.Infer<types.UnionType<Ts>> | null = null
+      for (const [variantName, variantType] of Object.entries(this.variants)) {
+        const concrete = types.concretise(variantType)
+        const result = concrete.decodeWithoutValidation(value, decodingOptions)
+        if (result.isOk) {
+          //look ahead with `validate` in order to get a correct variant if possible
+          const validateResult = concrete.validate(result.value as never)
+          if (validateResult.isOk) {
+            return result.map((v) => ({ [variantName]: v } as types.Infer<types.UnionType<Ts>>))
+          } else if (potentialDecoded === null) {
+            //keep this as potential variant but it will not validate
+            potentialDecoded = { [variantName]: result.value } as types.Infer<types.UnionType<Ts>>
+          }
+        } else {
+          errors.push(...result.error)
+        }
       }
+      if (potentialDecoded !== null) {
+        //returns the non validating variant
+        return decoding.succeed(potentialDecoded)
+      }
+      return decoding.failWithErrors(errors)
```

Comment:
I'm not sure but I think this could be rewritten with `result.tryEach`:

```suggestion
    } else {
      const entries = Object.entries(this.variants)
      const keepIfNotNull = (old, new) => old ? old : new
      const decodeVariant = ([variantName, variantType]) => types
        .concretise(variantType)
        .decodeWithoutValidation(value, options)
        .map((v) => ({ [variantName]: v }))
      
      return result.tryEach(entries, null, keepIfNotNull, [], mergeArrays, decodeVariant)
    }
```

I couldn't test this but it should work as expected, what do you think?

Reply:
I think it lacks the validation step and the `potentialDecoded` fallback.